### PR TITLE
Remove char_u (2): Replace char_u in remove_tail 

### PR DESCRIPTION
**Pull request by mbainter**
See: #459
Obsoletes: #2412

### CODE REVIEW

**Comment by elmart** [comment]
In commit message, it's `char_u`, not `u_char`.

File: src/nvim/os/env.c [hunk 1]
```diff
@@ -390,14 +390,14 @@ static char_u *vim_version_dir(char_u *vimdir)
 
 /// If the string between "p" and "pend" ends in "name/", return "pend" minus
 /// the length of "name/".  Otherwise return "pend".
-static char_u *remove_tail(char_u *p, char_u *pend, char_u *name)
+static char *remove_tail(char *p, char *pend, char *name)
 {
   int len = (int)STRLEN(name) + 1;
```
Comment (elmart) [inline src/nvim/os/env.c]: You could change len to `size_t` and remove cast.
That's unrelated to this PR's intent, so you can leave it out if you want.
But it costs so little in this case, that I'd do it.

File: src/nvim/os/env.c [hunk 2]
```diff
@@ -390,14 +390,14 @@ static char_u *vim_version_dir(char_u *vimdir)
...
-  char_u      *newend = pend - len;
+  char      *newend = pend - len;
```
Comment (elmart) [inline src/nvim/os/env.c]: You can take the chance to remove extra spaces.
We don't line up variables in new code.

File: src/nvim/os/env.c [hunk 3]
```diff
@@ -464,12 +464,14 @@ char_u *vim_getenv(char_u *name, bool *mustfree)
 
       /* remove "doc/" from 'helpfile', if present */
       if (p == p_hf)
-        pend = remove_tail(p, pend, (char_u *)"doc");
+        pend = (char_u *) remove_tail((char *)p, (char *)pend, (char *)"doc");
```
Comment (elmart) [inline src/nvim/os/env.c]: String literal "doc" doesn't need cast.

File: src/nvim/os/env.c [hunk 4]
```diff
@@ -464,12 +464,14 @@ char_u *vim_getenv(char_u *name, bool *mustfree)
...
 
       /* for $VIM, remove "runtime/" or "vim54/", if present */
       if (!vimruntime) {
-        pend = remove_tail(p, pend, (char_u *)RUNTIME_DIRNAME);
-        pend = remove_tail(p, pend, (char_u *)VIM_VERSION_NODOT);
+        pend = (char_u *) remove_tail((char *)p, (char *)pend,
+            (char *)RUNTIME_DIRNAME);
+        pend = (char_u *) remove_tail((char *)p, (char *)pend,
+            (char *)VIM_VERSION_NODOT);
```
Comment (elmart) [inline src/nvim/os/env.c]: String literals `RUNTIME_DIRNAME` and `VIM_VERSION_ROOT` don't need cast.

**Comment by elmart** [comment]
I think you're maybe taking the "small PR" a little bit too far.
Optimal granularity to me would be a PR with several commits like this (say 5), one per refactored function. But that's to save you some work. It's up to you.

File: src/nvim/os/env.c (hunk 2 above)
Comment (mbainter) [inline src/nvim/os/env.c]: Done

File: src/nvim/os/env.c (hunk 3 above)
Comment (mbainter) [inline src/nvim/os/env.c]: Sorry about that - I was running into a weird issue and did this as part of troubleshooting and forgot to roll it back.  Fixed.

**Comment by mbainter** [comment]
Got it.  I have a few more that depend on these two changes, and when I push them I'll do them together as individual commits in a single PR.

File: src/nvim/os/env.c (hunk 1 above)
Comment (mbainter) [inline src/nvim/os/env.c]: Done.

File: src/nvim/os/env.c [hunk 5]
```diff
@@ -464,12 +464,14 @@ char_u *vim_getenv(char_u *name, bool *mustfree)
 
       /* remove "doc/" from 'helpfile', if present */
       if (p == p_hf)
-        pend = remove_tail(p, pend, (char_u *)"doc");
+        pend = (char_u *) remove_tail((char *)p, (char *)pend, "doc");
```
Comment (elmart) [inline src/nvim/os/env.c]: Extra space after cast.

File: src/nvim/os/env.c [hunk 6]
```diff
@@ -464,12 +464,14 @@ char_u *vim_getenv(char_u *name, bool *mustfree)
...
 
       /* for $VIM, remove "runtime/" or "vim54/", if present */
       if (!vimruntime) {
-        pend = remove_tail(p, pend, (char_u *)RUNTIME_DIRNAME);
-        pend = remove_tail(p, pend, (char_u *)VIM_VERSION_NODOT);
+        pend = (char_u *) remove_tail((char *)p, (char *)pend,
```
Comment (elmart) [inline src/nvim/os/env.c]: Extra space after cast.

File: src/nvim/os/env.c [hunk 7]
```diff
@@ -464,12 +464,14 @@ char_u *vim_getenv(char_u *name, bool *mustfree)
...
+            RUNTIME_DIRNAME);
+        pend = (char_u *) remove_tail((char *)p, (char *)pend,
```
Comment (elmart) [inline src/nvim/os/env.c]: Extra space after cast.
I'm sorry I didn't see these before.

File: src/nvim/os/env.c (hunk 6 above)
Comment (elmart) [inline src/nvim/os/env.c]: After changes, it now fits in a single line.

File: src/nvim/os/env.c (hunk 7 above)
Comment (elmart) [inline src/nvim/os/env.c]: It doesn't fit in a single line yet. But you can lineup argument on separated line below first argument on line before. Our style guide suggests that.

**Comment by elmart** [comment]
I'm being extra-picky regarding style for you to quickly get out of common pitfalls that don't fit our style guide.

**Comment by splinterofchaos** [comment]
I looks like doing the same treatment for `vim_getenv()` would save us from introducing several temporary `char_u` casts. A quick grep of the source reveals that in only one instance do we not cast the `name` argument to `char_u` before calling.

**Comment by mbainter** [comment]
@splinterofchaos That's my goal.  I picked these fuctions because they're ones called from vim_getenv().  The next PR I give will fix the add_pathsep and after_pathsep functions, so I can clean up get_env.

I'm just breaking these down on the small side right now.  The next PR will be more complete, just with small commits instead.

**Comment by mbainter** [comment]
@elmart - Picking is welcome.  I should've thought to read the style-guide first.  I'll do that now.

**Comment by elmart** [comment]
LGTM.

**Comment by justinmk** [comment]
merged

**Comment by elmart** [comment]
@justinmk I was going to merge this and #2414. Didn't you see Assignee field? Of course, it's no problem. I'm saying this just in case you forgot about that, given that it's still a young practice.

**Comment by justinmk** [comment]
@elmart I'll try to start looking at the assignee field

**Comment by elmart** [comment]
@justinmk The Assignee field thing was agreed on with the main intention to help you, offloading some work from your shoulders. If you're not comfortable with it, please say so. Thanks!.

**Comment by justinmk** [comment]
@elmart I'm in favor of it, I just need to get used to it :)

File: src/nvim/os/env.c [hunk 8]
```diff
@@ -390,14 +390,14 @@ static char_u *vim_version_dir(char_u *vimdir)
 
 /// If the string between "p" and "pend" ends in "name/", return "pend" minus
 /// the length of "name/".  Otherwise return "pend".
-static char_u *remove_tail(char_u *p, char_u *pend, char_u *name)
+static char *remove_tail(char *p, char *pend, char *name)
 {
-  int len = (int)STRLEN(name) + 1;
-  char_u      *newend = pend - len;
+  size_t len = STRLEN(name) + 1;
```
Comment (philix) [inline src/nvim/os/env.c]: You can start replacing `STRLEN` with `strlen` when the string is `char *`. `STRLEN` is just a `strlen` wrapper that casts the string from `char_u *` to `char *`.

**Pull request closed without merging**